Use Prometheus instrumentation conventions

fabxc opened the pull request:
The `System` and `Subsystem` parameters are subject to removal.
(x-ref: https://github.com/prometheus/client_golang/issues/240)

All metrics should use base units, which is seconds in the duration
case.

Counters should always end in `_total` and metrics should avoid
referring to potential label dimensions. Those should rather be
mentioned in the documentation string.

@kubernetes/sig-instrumentation 

Reference docs:
https://prometheus.io/docs/practices/instrumentation/
https://prometheus.io/docs/practices/naming/

**Release note**:
```
Breaking change: Renamed REST client Prometheus metrics to follow the instrumentation conventions ("request_latency_microseconds" -> "rest_client_request_latency_seconds", "request_status_codes" -> "rest_client_requests_total"). Please update your alerting pipeline if you rely on them. 
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36704)
<!-- Reviewable:end -->

fabxc committed:
Use Prometheus instrumentation conventions

The `System` and `Subsystem` parameters are subject to removal.
(x-ref: https://github.com/prometheus/client_golang/issues/240)

All metrics should use base units, which is seconds in the duration
case.

Counters should always end in `_total` and metrics should avoid
referring to potential label dimensions. Those should rather be
mentioned in the documentation string.

grobie commented on pkg/client/metrics/prometheus/prometheus.go:
```diff
@@ -27,26 +27,22 @@ import (
-			Help:      "Request latency in microseconds. Broken down by verb and URL",
-			MaxAge:    time.Hour,
+			Name:   "rest_client_request_latency_seconds",
+			Help:   "Request latency in microseconds. Broken down by verb and URL.",
```
Seconds

grobie commented on pkg/client/metrics/prometheus/prometheus.go:
```diff
@@ -27,26 +27,22 @@ import (
-			Name:      "request_status_codes",
-			Help:      "Number of http requests, partitioned by metadata",
+			Name: "rest_client_requests_total",
+			Help: "Number of HTTP requests, partitioned by metadata.",
```
Metadata is not descriptive. Name the two labels.

grobie commented on pkg/client/metrics/prometheus/prometheus.go:
```diff
@@ -27,26 +27,22 @@ import (
 var (
 	// requestLatency is a Prometheus Summary metric type partitioned by
 	// "verb" and "url" labels. It is used for the rest client latency metrics.
 	requestLatency = prometheus.NewSummaryVec(
```
As we can't aggregate quantiles, wouldn't it be a lot more helpful to use a histogram here?

fabxc replied:
I considered it but as a default it might not be ideal:
* there are quite a few possible paths and histograms generate a fair amount of metrics – so unexpected blow-up
* we have to agree on a single bucketing pattern fitting latencies in all setups, which is not really possible – otherwise, people can just specify their own metrics in the first place
* I don't expect most clients to be part of a horizontally scaling service but rather controllers/operators/etc., which don't profit too much from histograms

grobie commented:
Given there will be usually only a few controller instances, is the
increased number of time series an actual problem?

The use case is not so much to aggregate across controllers, but to
aggregate the latencies of a single instance to answer questions like the
overall client latency. Due to the current bucketing, an operator always
has to plot all summaries for all labels to spot issues. And then you still
don't know whether that single outlier was just due to a single request
(and staleness) or is a real issue.

I found the summaries in this situation very impractical back at
SoundCloud. @matthiasr

On Sun, Nov 13, 2016, 11:29 Fabian Reinartz <notifications@github.com>
wrote:

> *@fabxc* commented on this pull request.
> ------------------------------
>
> In pkg/client/metrics/prometheus/prometheus.go
> <https://github.com/kubernetes/kubernetes/pull/36704>:
>
> > @@ -27,26 +27,22 @@ import (
>  	"github.com/prometheus/client_golang/prometheus"
>  )
>
> -const restClientSubsystem = "rest_client"
> -
>  var (
>  	// requestLatency is a Prometheus Summary metric type partitioned by
>  	// "verb" and "url" labels. It is used for the rest client latency metrics.
>  	requestLatency = prometheus.NewSummaryVec(
>
> I considered it but as a default it might not be ideal:
>
>    - there are quite a few possible paths and histograms generate a fair
>    amount of metrics – so unexpected blow-up
>    - we have to agree on a single bucketing pattern fitting latencies in
>    all setups, which is not really possible – otherwise, people can just
>    specify their own metrics in the first place
>    - I don't expect most clients to be part of a horizontally scaling
>    service but rather controllers/operators/etc., which don't profit too much
>    from histograms
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/36704>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAANaNE2Z4rhgu-oRmEZF0-hYkm_Kjbpks5q90kXgaJpZM4KwfB2>
> .
>

piosz commented:
cc @gmarek @wojtek-t

gmarek commented:
cc @Q-Lee - we'll need to do something with our metrics.

fabxc committed:
Use Histogram instead of Summary

A histogram allows to aggregate by labels and calculate more
comprehensive quantiles.

fabxc commented:
@grobie  I see the point to aggregate internally. Updated to use histograms – in the end, it's just a few lines people can easily copy&adapt as it fits their application.

gmarek commented:
@fgrzadkowski

deads2k commented:
Seconds seem needlessly coarse, but I can't say that I feel strongly.

@smarterclayton bother you one way or the other?

fabxc commented:
Prometheus has no first-class concept of units (for better or worse (probably better)) so taking the simplest (shorter name, easier code as in this case, ...) base unit allows us to do operations between different metrics without having to care about unit conversions.
We just apply a single conversion to the final result for human readability.

fgrzadkowski commented on pkg/client/metrics/prometheus/prometheus.go:
```diff
@@ -27,26 +27,22 @@ import (
-			Name:      "request_status_codes",
-			Help:      "Number of http requests, partitioned by metadata",
+			Name: "rest_client_requests_total",
+			Help: "Number of HTTP requests, partitioned by status code, method, and host.",
```
@Q-Lee Do you know if we are using those metrics for our monitoring?

fgrzadkowski commented:
This seems like a breaking change for those who would use these metrics for monitoring purposes.

First, we should be very careful with this change and secondly we need to release note label, that would properly describe it.

fabxc commented:
@fgrzadkowski yes, what is the policy on breaking metrics for Kubernetes?

For Prometheus 1.x we explicitly excluded its own metrics from the stability guarantees. One reason is at least some metrics are about application internals, which may be implementation specific, which may obviously change at any point in time.
For etcd we ultimately wanted both, some stable metrics which are frequently used in alerting (usually "symptom metrics") and dashboards, and some flexible ones investigation (usualy "cause metrics"). They are prefixed by `etcd_` and `etcd_debugging` respectively and the former are guaranteed to be stable. Something similar may make sense for Kubernetes eventually too.

For Kubernetes in general there are quite a few more places that would have to get aligned with best practices before though.

fgrzadkowski commented:
I think it's ok to have a breaking change here, but it should well documented in release nodes.

@Q-Lee

piosz commented:
@fabxc could you please add release notes to the PR and then it seems it's ok to merge.

fabxc commented:
@piosz sorry about the delay. Is the release note sufficient?

piosz commented:
@fabxc I updated the release note a little bit. Please take w look whether you are ok with them.

piosz commented:
@deads2k could you please approve?

deads2k commented:
/approve

piosz commented:
Closed/reopened to trigger tests

fabxc commented:
Missed that notification. 👍

smarterclayton commented:
This should have made it into 1.6 - it's been sitting in the queue for over a month.

@k8s-bot test this issue #IGNORE

calebamiles commented:
@k8s-bot cvm gce e2e test this

calebamiles commented:
Added this issue to the 1.6 milestone, @fabxc, @smarterclayton. Please be responsive to test failures if the intention is to have this PR merged for 1.6. Thanks!

**cc: @ethernetdan, @kubernetes/release-team**

fabxc commented:
I couldn't connect any of the test failures up until now to the changes I've made.
I rebased once again in case that helps.

deads2k commented:
@k8s-bot cvm gce e2e test this
